src/zftools/metagraph.pyx

```diff
@@ -31,15 +31,35 @@ from cpython.mem cimport (
 from zftools.fastqueue cimport FastQueueForBFS
 
 
+class GraphIsDirectedError(NotImplementedError):
+    pass
+
+class GraphAllowsLoopsError(NotImplementedError):
+    pass
+
+
+
+
 def zero_forcing_set(sage_graph):
     cdef:
-        ZFSearchMetagraph metagraph = ZFSearchMetagraph(sage_graph)
-        frozenset start = frozenset()
-        frozenset end = frozenset(
-                metagraph.to_relabeled_metavertex(sage_graph.vertices(sort=False))
-                )
-
-    return metagraph.dijkstra(start, end)
+        ZFSearchMetagraph metagraph
+        frozenset start
+        frozenset end
+    
+    try:
```

Comment:
Now that I see the end result, we should actually just let the exception bubble up here, that is, not do any try/catching. The reason for that is, as the code currently is, all we do is print a message. If someone calls this function from deep in some codebase, the print won't help them too much as the function call will just (likely, I haven't tested) do nothing and return `None`, and they might have to dig very deep to see why their code isn't working.

We should put those error messages in the exceptions themselves in `ZFSearchMetagraph.__init__` — `raise GraphIsDirectedError("This is a...")`. This way, the code becomes cleaner, and the user's code will crash if the exception is not caught, forcing them to handle it however they see fit.

Reply:
Your comment about what happens if “someone calls this function from deep in some codebase” makes total sense.  It is certainly possible that someone would use `zero_forcing_set()` to implement the computation of some other thing, at least in theory.  And so, while I was thinking of `zero_forcing_set()` as being _the_ interface to what we have working here, that is probably short-sighted.  All that is to say, we can probably do better than some dumb `print` statements.

At the same time, I think the reason not to _just_ “let the exception bubble up” is that `zero_forcing_set()` is the place where we should be able to instantiate different subclasses of the metagraph class according to what kind of forcing is called for.  We could either (a) have `zero_forcing_set()` itself sort out what kind of graph it is dealing with and throw exceptions if it can’t handle them or (b) code each subclass so that it can cough up an exception and then branch on why it’s complaining that it cannot handle forcing on the graph we are handing it.

Any ideas on which of (a) or (b) is better, or if there is an option (c) that I am not considering?

Reply:
I think (b) is the better option personally, because with (a), if someone were to manually create a `ZFSearchMetagraph` with a directed graph, they would presumably get an answer, but an incorrect one. (b) allows the class to have fine-grained control over what it can and cannot do. Sure, the user isn't _supposed_ to create it manually, but it's better to be safe than sorry IMO.

I suppose we can leave the `try/except` in `zero_forcing_set()` for now to reinforce that there will eventually be cases that will be handled. Regarding that though, do we know how exactly we're going to structure these new subclasses? Will we have a version of `ZFSearchMetagraph` that works on directed graphs? If so, I presume this will be its only functionality? If that is indeed the case, that means we will likely have to make 2 separate subclasses for directed graphs vs. undirected graphs for each new algorithm (i.e., the positive semidefinite ZF number), and 2 more for looped vs. non-looped graphs, right? If all of that sounds right, there must be a better way, as that doesn't really pass the sniff test for me.

Regarding what I meant for my previous comments, we're close, but let me push what I had in mind. You can certainly revert or write over it if you'd like also.

Reply:
In the commit I just pushed (https://github.com/alexhutman/zftools/pull/19/commits/e91a6e84377d6b7386235de6b31a06fca1f9c872), I just added the error message to the actual creation of the exception. I also removed the `try/catch` for now -- if you want, you can revert that part and do something like:
```python3
cdef:
    ZFSearchMetagraph metagraph
    frozenset start
    frozenset end
   
try:
    metagraph = ZFSearchMetagraph(sage_graph)
except GraphIsDirectedError:
    # This is where we will instantiate a metagraph for directed forcing
    raise
except GraphAllowsLoopsError:
    # This is where we will instantiate a metagraph for (undirected) looped forcing
    raise
start = frozenset()
end = frozenset(metagraph.to_relabeled_metavertex(sage_graph.vertices(sort=False)))
return metagraph.dijkstra(start, end)
```

In my opinion, this is a little unnecessary since we can just add this once we get closer to actually handling these exceptions via the new subclasses, but I wouldn't be opposed to it. [Here](https://docs.python.org/3/tutorial/errors.html#raising-exceptions) is a section from the Python docs for reference. Note the bottom of the last example in the section: we can call `raise` (with no expression afterwards) inside of an `except` clause in order to re-raise the caught exception.